Hopefully improve performance involving synchronization

noobanidus opened the pull request:
By only ever writing to the pending entries, this should prevent off-thread modification of the main block entities list which previously triggered CMEs; this hopefully removes the need to copy the incoming list (which, with large lists, can be very expensive).

With that out of the way, switch to using an Iterator's remove function instead of maintaining a "to remove" list.

It might be possible to entirely remove the world lock.

Convert `Entry` into a record, although maintain the dimension/block position hashing and equality checks to ensure duplicates are properly handled.

Finally, improve the "loaded chunks" handler: it's annoying that this exists, but in all of the testing and the releases that I previously made, `ChunkSource::hasChunk` was never sufficient to determine if a chunk was *actually* fully loaded or not.

In theory, ConcurrentHashMap should only be locking on relevant buckets, although in this case those buckets are most likely to be `ResourceKey<Level>(minecraft:overworld)`. It may be worth having dedicated `ConcurrentHashMap.newKeySet`s for each dimension instead.

noobanidus committed:
Test improving chunk loading performance.

noobanidus committed:
Test improving block entity conversion lag.

noobanidus commented on fabric/src/main/java/noobanidus/mods/lootr/fabric/event/HandleChunk.java:
```diff
@@ -1,35 +1,30 @@
 
 public class HandleChunk {
-  public static final Map<ResourceKey<Level>, Set<ChunkPos>> LOADED_CHUNKS = Collections.synchronizedMap(new Object2ObjectLinkedOpenHashMap<>());
+  public static final Map<ResourceKey<Level>, Set<ChunkPos>> LOADED_CHUNKS = new ConcurrentHashMap<>();
```
Functionally this doesn't seem any different than just the previous `synchronized`, as it will only synchronize on buckets (generally speaking), but if those buckets are all the overworld?

chylex replied:
For `ConcurrentHashMap`, read operations do not lock at all, even when the map is being updated concurrently - the reading thread will simply not see the change until the updating thread finishes it.

So even if the map only contains one Level, it will be way better than a synchronized map for `LootrAPIImpl#anyUnloadedChunks`, which only calls `LOADED_CHUNKS.get` and therefore doesn't lock.

mezz replied:
You only do one operation on a time on this map, so it makes sense to me to use a `ConcurrentHashMap` and feel safe. There isn't a way to squeeze more performance out of this using `synchronized` unless you can batch multiple updates at once somehow.

noobanidus commented on fabric/src/main/java/noobanidus/mods/lootr/fabric/event/HandleChunk.java:
```diff
@@ -1,35 +1,30 @@
-          Set<ChunkPos> chunkSet = LOADED_CHUNKS.computeIfAbsent(chunk.getLevel().dimension(), k -> Collections.synchronizedSet(new ObjectLinkedOpenHashSet<>()));
-          chunkSet.add(chunk.getPos());
-        }
+        Set<ChunkPos> chunkSet = LOADED_CHUNKS.computeIfAbsent(chunk.getLevel()
```
Once again I'm unable to determine if `computeIfAbsent` is really safe.

chylex replied:
For JDK `Map` implementations, `computeIfAbsent` behaves atomically, so **it is safe**, but here's some more info if you're curious:

The default implementation of `computeIfAbsent` is a sequence of `get-compare-put`, which requires special treatment for thread-safe maps, otherwise there would be a race condition:
1. Two threads call get and both find no value.
2. Two threads both compute the new value.
3. Two threads both put the new value, one overwriting the other. Then, `computeIfAbsent` would return a different instance of `Set<ChunkPos>` in each thread, and one of the subsequent `Set.add` operations would get lost.

Both `ConcurrentHashMap` and `SynchronizedMap` (`Collections.synchronizedMap`) override `computeIfAbsent`, in a way that guarantees that the value will only be computed and put once, no matter how many threads are calling it at the same time.

Thread-safe `Map` implementations, from third-party libraries that were written before Java 8, inherit the default implementation of `computeIfAbsent`, so even though the individual `get` and `put` operations may be synchronized and therefore there is no risk of corrupting the internal data structures, the `computeIfAbsent` operation as a whole would not be synchronized, and you would risk the race condition I described.

mezz replied:
`ConcurrentHashMap` directly implements `computeIfAbsent` so that it is concurrency-safe

noobanidus commented on common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java:
```diff
@@ -72,35 +72,26 @@ public static void addEntry(RandomizableContainerBlockEntity incoming, Level lev
-      } else {
-        blockEntityEntries.add(newEntry);
-      }
+      pendingEntries.add(newEntry);
```
This means `pendingEntries` is only ever modified within the context of `listLock`.

This allows us to only ever modify `blockEntityEntries` while on the main thread.

noobanidus commented on common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java:
```diff
@@ -72,35 +72,26 @@ public static void addEntry(RandomizableContainerBlockEntity incoming, Level lev
-      } else {
-        blockEntityEntries.add(newEntry);
-      }
+      pendingEntries.add(newEntry);
```
Would it be worth converting the pending entries to a `ConcurrentHashMap` and using an `Iterator` to both `add` to the next cycle of block entities, and `remove` later to avoid locking?

mezz replied:
The locking you have is extremely brief and you benefit from locking over a large operations and avoiding individual locks on iteration.
I think your approach is better than `ConcurrentHashMap` but to be more confident, you can try to benchmark it.

noobanidus commented on common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java:
```diff
@@ -72,35 +72,26 @@ public static void addEntry(RandomizableContainerBlockEntity incoming, Level lev
-      tickingList = true;
-      copy = new ObjectOpenHashSet<>(blockEntityEntries);
-      tickingList = false;
-    }
```
In instances where a *huge* number of semi-eligible block entities (specifically Lost Cities with dispensers and droppers), this copy operation can quickly become the most expensive thing.

Switching to `Iterator::remove` would seem to bypass this issue.

mezz replied:
working to eliminate the copy is the right idea

noobanidus commented on common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java:
```diff
@@ -167,44 +159,29 @@ public static void onServerTick() {
-      blockEntityEntries.removeAll(toRemove);
       blockEntityEntries.addAll(pendingEntries);
-      tickingList = false;
       pendingEntries.clear();
```
As mentioned earlier, it might be worth it to use a `ConcurrentHashMap` for `pendingEntries` instead.

chylex replied:
If you're not seeing a performance issue, I would leave it like this.

Otherwise it gets tricky, because you would need to drain the `pendingEntries` atomically, to avoid losing an entry if one thread called `pendingEntries.add`, while another thread was exactly between `blockEntityEntries.addAll(pendingEntries)` and `pendingEntries.clear()`.

noobanidus commented on common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java:
```diff
@@ -72,35 +72,26 @@ public static void addEntry(RandomizableContainerBlockEntity incoming, Level lev
-      copy = new ObjectOpenHashSet<>(blockEntityEntries);
-      tickingList = false;
-    }
     synchronized (worldLock) {
```
I'm not even sure if this section *needs* to be synchronized to anything.

chylex replied:
Afaik `onServerTick` can only called from one thread, so the `worldLock` is pointless.

mezz replied:
This section doesn't need to by synchronized, because ticks always happen one at a time on the same thread.

To be paranoid, you could make the entire method synchronized
```java
public static synchronized void onServerTick() {
```
and either way it will be optimized to almost no overhead after a little while.

However, trying to defend against multiple game ticks happening concurrently may be just protecting your precious code while the rest of the game is being completely annihilated. I might do it just to prevent getting any crash reports but it's up to you :D

mezz commented on common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java:
```diff
@@ -72,35 +72,26 @@ public static void addEntry(RandomizableContainerBlockEntity incoming, Level lev
 
     Entry newEntry = new Entry(dimension, position, chunks, LootrAPI.getCurrentTicks());
     synchronized (listLock) {
```
There is really only one list that needs protection in this class, so I recommend making this 
```suggestion
    synchronized (pendingEntries) {
```
and removing `listLock`

mezz commented on fabric/src/main/java/noobanidus/mods/lootr/fabric/impl/LootrAPIImpl.java:
```diff
@@ -348,21 +346,22 @@ public boolean performPiecewiseCheck() {
-        }
+    for (ChunkPos myPos : chunks) {
+      if (!syncedChunks.contains(myPos)) {
+        return true;
```
⚠️ this is only safe as long as `syncedChunks` is concurrent-safe. We know it is, but by having the logic here we're leaking abstractions outside of the `HandleChunk` class.

I recommend making a public method in `HandleChunk` that handles this logic, and do not make `HandleChunk.LOADED_CHUNKS` public, it's too dangerous. Then all your concurrency concerns can be hidden inside `HandleChunk`.

noobanidus committed:
Address concurrency leaks with "HandleChunk".

Centralize all the logic in the `HandleChunk` (common) class and defer
to that for adding new chunks and checking for "any unloaded" chunks,
allowing the `LOADED_CHUNKS` to always be private.

noobanidus committed:
Remove the world lock.

The server tick always happens on the main thread, so this
synchronization is pointless.

noobanidus committed:
Remove listLock and synchronize on `pendingEntries`.

noobanidus committed:
Switch to an ObjectOpenHashSet from linked.

noobanidus committed:
Fix "missing loot table" logic.

This is only supposed to check for blacklisted loot tables. It's
possible, at this point, for the loot table to not have been set yet.

(I don't appear to have back-ported the aggressive mode to 1.21 yet.)

noobanidus committed:
Move world border check to BlockEntityTicker::addEntry.

noobanidus committed:
Move blacklisted loot table check earlier.

Micro-optimization to avoid constructing the "potentially outside of the
world border" chunks.

noobanidus committed:
"Explain" some of the checking.

mezz committed:
improvements for BlockEntityTicker

mezz commented:
@noobanidus I messed around with this more in a branch, here:
* https://github.com/LootrMinecraft/Lootr/pull/671

mezz committed:
fix build

mezz committed:
wip

mezz committed:
remove LoadedChunks cache

mezz committed:
improve tick performance

mezz committed:
Merge remote-tracking branch 'origin/arch-1.21-test' into arch-1.21-test

# Conflicts:
#	common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java
#	common/src/main/java/noobanidus/mods/lootr/common/mixins/MixinLevelChunk.java
#	fabric/src/main/java/noobanidus/mods/lootr/fabric/event/LootrEventsInit.java
#	neoforge/src/main/java/noobanidus/mods/lootr/neoforge/event/HandleChunkImpl.java
#	neoforge/src/main/java/noobanidus/mods/lootr/neoforge/impl/LootrAPIImpl.java

mezz committed:
try to simplify diff

mezz committed:
reduce diff

mezz committed:
check full chunk load status, make sure it's ticking

mezz committed:
go back to safer entity detection method

mezz committed:
split tickers by level

mezz committed:
fix ticking check

mezz committed:
Re-implement the LoadedChunks cache

mezz committed:
fix case where loaded chunks might be null

mezz committed:
fix comment

mezz committed:
add warning log to check for containers with no loot table

noobanidus committed:
Merge pull request #671 from mezz/arch-1.21-test

Improve performance involving synchronization
